go/arrow/array/diff.go

```diff
@@ -0,0 +1,276 @@
+// elements but both fields of the struct must have the same length. To accommodate this
+// the first element of "insert" should be ignored.)
+//
+// For example for base "hlloo" and target "hello", the edit script would be
+// [
+//
+//	{"insert": false, "run_length": 1}, // leading run of length 1 ("h")
+//	{"insert": true, "run_length": 3}, // insert("e") then a run of length 3 ("llo")
+//	{"insert": false, "run_length": 0} // delete("o") then an empty run
+//
+// ]
+//
+// base: baseline for comparison
+// target: an array of identical type to base whose elements differ from base's
+// mem: memory to store the result will be allocated from this memory pool
+func Diff(base, target arrow.Array, mem memory.Allocator) (*Struct, error) {
+	if base.DataType().Fingerprint() != target.DataType().Fingerprint() {
+		return nil, errors.New("only taking the diff of like-typed arrays is supported")
+	}
+	switch base.DataType().ID() {
+	case arrow.EXTENSION:
+		return Diff(base.(ExtensionArray).Storage(), target.(ExtensionArray).Storage(), mem)
+	case arrow.DICTIONARY:
+		return nil, fmt.Errorf("diffing arrays of type %s is not implemented", base.DataType().String())
+	case arrow.RUN_END_ENCODED:
+		return nil, fmt.Errorf("diffing arrays of type %s is not implemented", base.DataType().String())
+	}
+	d := newQuadraticSpaceMyersDiff(base, target)
+	r, err := d.Diff(mem)
+	if err != nil {
+		if r != nil {
+			r.Release()
+		}
+		return nil, err
+	}
+	return r, nil
+}
+
+// editPoint represents an intermediate state in the comparison of two arrays
+type editPoint struct {
+	base   int
+	target int
+}
+
+func (e editPoint) Equals(other editPoint) bool {
+	return e.base == other.base && e.target == other.target
+}
+
+type quadraticSpaceMyersDiff struct {
+	base         arrow.Array
+	target       arrow.Array
+	finishIndex  int
+	editCount    int
+	endpointBase []int
+	insert       []bool
+	baseBegin    int
+	targetBegin  int
+	baseEnd      int
+	targetEnd    int
+}
+
+func newQuadraticSpaceMyersDiff(base, target arrow.Array) *quadraticSpaceMyersDiff {
+	d := &quadraticSpaceMyersDiff{
+		base:         base,
+		target:       target,
+		finishIndex:  -1,
+		editCount:    0,
+		endpointBase: []int{},
+		insert:       []bool{},
+		baseBegin:    0,
+		targetBegin:  0,
+		baseEnd:      base.Len(),
+		targetEnd:    target.Len(),
+	}
+	d.endpointBase = []int{d.extendFrom(editPoint{d.baseBegin, d.targetBegin}).base}
+	if d.baseEnd-d.baseBegin == d.targetEnd-d.targetBegin && d.endpointBase[0] == d.baseEnd {
+		// trivial case: base == target
+		d.finishIndex = 0
+	}
+	return d
+}
+
+func (d *quadraticSpaceMyersDiff) valuesEqual(baseIndex, targetIndex int) bool {
+	baseNull := d.base.IsNull(baseIndex)
+	targetNull := d.target.IsNull(targetIndex)
+	if baseNull || targetNull {
+		return baseNull && targetNull
+	}
+	return SliceEqual(d.base, int64(baseIndex), int64(baseIndex+1), d.target, int64(targetIndex), int64(targetIndex+1))
```

Comment:
yea, i can see where an `UntypedValue` method would potentially be a more efficient alternative here, but given the current things available this works.